design/build-cache.md

```diff
@@ -0,0 +1,201 @@
+
+## Status
+
+- 29 June 2018: Proposal submitted
+
+## Proposal
+
+Provide a mechanism for transparent caching of resources between jobs within a single pipeline, or between consecutive executions of a particular pipeline.
+
+1. Content to be cached can be specified by the user.
+2. Caching is scoped. 
+   1. Job level caching
+   2. Event level caching
+   3. Pipeline level caching (static caching, in a sense)
+3. Well used caching patterns can be offered as predefined options.
+   1. Dependency caching for build systems such as npm, sbt and cargo.
+
+The build cache and cache usage is configurable via Yaml and backed by an internal sd-command and caching service.
+
+## Details
+
+### Yaml specification
+
+```yml
+shared:
+  # Caches defined in the shared section can be pipeline or exection scoped.
+  cache:
+    # A pipeline scoped cache.
+    pipeline:  
+      # The caching targets
+      target: ["build/**", "~/.sbt"]
+    # An event scoped cache
+    event:
+      target: ["target/deployment.jar"]
+    # A job scoped cache 
+    my-cache-1:
+      target: ["build/some-artifact.zip"]
+jobs:
+  test:
+    requires: [~pr]
+    image: rust:latest
+    steps:
+      - test: cargo test
+  build: 
+    requires: [~pr]
+    # Indicate which job scoped caches to use.
+    cache: ["my-cache-1"]
+    image: rust:latest
+    steps:
+      - test: cargo build
+  publish-preview:
+    # Indicate which job scoped caches to use.
+    cache: ["my-cache-1"]
+    requires: [~commit]
+    image: rust:latest
+    steps:
+      - add_musl: rustup target install x86_64-unknown-linux-musl
+      - build: cargo build --release --target=x86_64-unknown-linux-musl
+```
+
+### Design considerations
+
+1. Resource usage
+   1. Copying to and from a remote server for every job in a pipeline can lead to unnecessary network bandwidth usage and unnecessarily slow speeds.
+2. Integrity and consistency of cached data.
+3. User experience
+   1. Is it easily adaptable to new use cases?
+   2. Is it easily enabled and administered?
+   3. Is it easily configurable by end users once enabled?
+   4. Does it speed up build time? How much does it speed up build time for various use cases?
+   5. Is it consistent?
+   6. Is it reliable?
+4. Security
+   1. Is it possible to inject malicious code into pipelines via the cache mechanism? Malicious pull requests are a good example of this.
+
+### Security
+
+Injecting malicious code into a build via the caching mechanism is a real concern. This can lead to theft of tokens or credentials from job, or in even lead to malicious code being deployed to a production environment. 
+
+Pull requests, as the main vector for injection of malicious code, require additional security considerations. Pull requests therefore treat caches as readonly. 
+
+Caching of potentially sensitive data also needs to be handled. Data provided via the secrets field should not be cached. 
```

Comment:
Testing for this might be expensive/troublesome. A simple `grep` could turn out to be quite expensive. Any suggestions for this?

Reply:
There might not be security issue if we only limit read access of the cache content to the particular pipeline?

Also maybe we can defer the responsibility of not putting sensitive data to users for now.

Reply:
@Zhongtang @tk3fftk @catto Any thoughts on security issues?